app/services/invoices/payments/cashfree_service.rb

```diff
@@ -0,0 +1,202 @@
+
+        unless invoice.total_amount_cents.positive?
+          update_invoice_payment_status(payment_status: :succeeded)
+          return result
+        end
+
+        Rails.logger.debug "invoice.total_amount_cents 2: #{invoice.total_amount_cents}"
+
+        increment_payment_attempts
+
+        payment = Payment.new(
+          payable: invoice,
+          payment_provider_id: cashfree_payment_provider.id,
+          payment_provider_customer_id: customer.cashfree_customer.id,
+          amount_cents: invoice.total_amount_cents,
+          amount_currency: invoice.currency.upcase,
+          provider_payment_id: invoice.id,
+          status: :pending
+        )
+        payment.save!
+
+        result.payment = payment
+        result
+      end
+
+      def update_payment_status(provider_payment_id:, status:, metadata: {})
+        payment = if metadata[:payment_type] == 'one-time'
+          create_payment(provider_payment_id:)
+        else
+          Payment.find_by(provider_payment_id:)
+        end
+        return result.not_found_failure!(resource: 'cashfree_payment') unless payment
+
+        result.payment = payment
+        result.invoice = payment.payable
+        return result if payment.payable.payment_succeeded?
+
+        invoice_payment_status = invoice_payment_status(status)
+
+        payment.update!(status: invoice_payment_status)
+        update_invoice_payment_status(payment_status: invoice_payment_status)
+
+        result
+      rescue BaseService::FailedResult => e
+        result.fail_with_error!(e)
+      end
+
+      def generate_payment_url
+        return result unless should_process_payment?
+
+        res = create_post_request(payment_url_params)
+
+        result.payment_url = JSON.parse(res.body)["link_url"]
+
+        result
+      rescue LagoHttpClient::HttpError => e
+        deliver_error_webhook(e)
+        result.service_failure!(code: e.error_code, message: e.error_body)
+      end
+
+      private
+
+      attr_accessor :invoice
+
+      delegate :organization, :customer, to: :invoice
+
+      def create_payment(provider_payment_id:)
+        @invoice = Invoice.find_by(id: provider_payment_id)
+
+        increment_payment_attempts
+
+        Payment.new(
+          payable: invoice,
+          payment_provider_id: cashfree_payment_provider.id,
+          payment_provider_customer_id: customer.cashfree_customer.id,
+          amount_cents: invoice.total_amount_cents,
+          amount_currency: invoice.currency.upcase,
+          provider_payment_id:
+        )
+      end
+
+      def should_process_payment?
+        return false if invoice.payment_succeeded? || invoice.voided?
+        return false if cashfree_payment_provider.blank?
+
+        customer&.cashfree_customer&.id
+      end
+
+      def client
+        @client ||= LagoHttpClient::Client.new(::PaymentProviders::CashfreeProvider::BASE_URL)
+      end
+
+      def create_post_request(body)
+        client.post_with_response(body, {
```

Comment:
It looks like you have to initialize the client with the service URL somewhere

Reply:
Could you please explain this? I didn't understand it.

Reply:
Of sorry, I missed the method right before this one. Ignore this comment, everything looks fine here :)